File: citycompass/lib/schemas/survey.ts
import { z } from "zod";

// Step 1: Personal Information
export const personalInfoSchema = z.object({
  name: z.string().min(2, "Name must be at least 2 characters"),
  email: z.string().email("Please enter a valid email address"),
  age: z.number().min(18, "You must be at least 18 years old").max(100, "Age must be less than 100"),
  borough: z.enum(["Manhattan", "Brooklyn", "Queens", "Bronx", "Staten Island"], {
    message: "Please select a borough",
  }),
  neighborhood: z.string().min(2, "Please enter your neighborhood"),
  zipCode: z.string().regex(/^\d{5}$/, "Please enter a valid 5-digit zip code"),
  residencyDuration: z.enum(["< 6 months", "6-12 months", "1-3 years", "3-5 years", "5+ years"], {
    message: "Please select your residency duration",
  }),
});

// Step 2: Safety & Security
export const safetySecuritySchema = z.object({
  safetyRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  timeOfDaySafety: z.enum(["Day only", "Night only", "Both", "Neither"], {
    message: "Please select when you feel safe",
  }),
  crimeConcernLevel: z.enum(["Not concerned", "Slightly concerned", "Moderately concerned", "Very concerned"], {
    message: "Please select your concern level",
  }),
  policePresenceRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  safetyTestimonial: z.string().optional(),
});

// Step 3: Cleanliness & Environment
export const cleanlinessEnvironmentSchema = z.object({
  streetCleanlinessRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  trashManagementRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  parksQualityRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  noiseLevel: z.enum(["Very quiet", "Quiet", "Moderate", "Loud", "Very loud"], {
    message: "Please select the noise level",
  }),
  environmentalTestimonial: z.string().optional(),
});

// Step 4: Food & Amenities
export const foodAmenitiesSchema = z.object({
  groceryStoreAccess: z.enum(["Walk < 5min", "Walk 5-15min", "Walk 15-30min", "Drive required", "No access"], {
    message: "Please select grocery store access",
  }),
  restaurantVarietyRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  foodAffordabilityRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  farmersMarketAccess: z.boolean(),
  foodAccessTestimonial: z.string().optional(),
});

// Step 5: Financials & Living Costs
export const financialsSchema = z.object({
  rentAffordability: z.enum(["Very affordable", "Affordable", "Fair", "Expensive", "Very expensive"], {
    message: "Please select affordability level",
  }),
  costOfLiving: z.enum(["Very low", "Low", "Moderate", "High", "Very high"], {
    message: "Please select cost of living level",
  }),
  valueForMoneyRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  financialTestimonial: z.string().optional(),
});

// ... rest of the file stays the same ...

// Step 6: Overall Feedback
export const overallFeedbackSchema = z.object({
  overallRating: z.number().min(1, "Please select a rating").max(5, "Rating must be between 1 and 5"),
  wouldRecommend: z.boolean(),
  biggestStrength: z.string().min(10, "Please provide at least 10 characters"),
  areaForImprovement: z.string().min(10, "Please provide at least 10 characters"),
  additionalComments: z.string().optional(),
});

// Complete survey form schema
export const surveyFormSchema = personalInfoSchema
  .merge(safetySecuritySchema)
  .merge(cleanlinessEnvironmentSchema)
  .merge(foodAmenitiesSchema)
  .merge(financialsSchema)
  .merge(overallFeedbackSchema);

export type PersonalInfoData = z.infer<typeof personalInfoSchema>;
export type SafetySecurityData = z.infer<typeof safetySecuritySchema>;
export type CleanlinessEnvironmentData = z.infer<typeof cleanlinessEnvironmentSchema>;
export type FoodAmenitiesData = z.infer<typeof foodAmenitiesSchema>;
export type FinancialsData = z.infer<typeof financialsSchema>;
export type OverallFeedbackData = z.infer<typeof overallFeedbackSchema>;
export type SurveyFormData = z.infer<typeof surveyFormSchema>;

File: backend/app/schemas/survey.py
from pydantic import BaseModel, EmailStr, Field, field_validator
from typing import Optional, Literal
from datetime import datetime
import re


# Enums as Literal types for strict validation
BoroughType = Literal["Manhattan", "Brooklyn", "Queens", "Bronx", "Staten Island"]
ResidencyDurationType = Literal["< 6 months", "6-12 months", "1-3 years", "3-5 years", "5+ years"]
TimeOfDaySafetyType = Literal["Day only", "Night only", "Both", "Neither"]
CrimeConcernType = Literal["Not concerned", "Slightly concerned", "Moderately concerned", "Very concerned"]
NoiseLevelType = Literal["Very quiet", "Quiet", "Moderate", "Loud", "Very loud"]
GroceryAccessType = Literal["Walk < 5min", "Walk 5-15min", "Walk 15-30min", "Drive required", "No access"]
AffordabilityType = Literal["Very affordable", "Affordable", "Fair", "Expensive", "Very expensive"]
CostOfLivingType = Literal["Very low", "Low", "Moderate", "High", "Very high"]


class SurveyRequest(BaseModel):
    """Survey request schema with full validation matching frontend"""
    
    # Step 1: Personal Information
    name: str = Field(..., min_length=2, description="Full name of the respondent")
    age: int = Field(..., ge=18, le=100, description="Age must be between 18 and 100")
    borough: BoroughType = Field(..., description="NYC Borough")
    neighborhood: str = Field(..., min_length=2, description="Neighborhood name")
    zipCode: str = Field(..., description="5-digit zip code")
    residencyDuration: ResidencyDurationType = Field(..., description="How long lived in neighborhood")
    
    # Step 2: Safety & Security
    safetyRating: int = Field(..., ge=1, le=5, description="Overall safety rating 1-5")
    timeOfDaySafety: TimeOfDaySafetyType = Field(..., description="When they feel safe")
    crimeConcernLevel: CrimeConcernType = Field(..., description="Level of crime concern")
    policePresenceRating: int = Field(..., ge=1, le=5, description="Police presence rating 1-5")
    safetyTestimonial: Optional[str] = Field(None, description="Optional safety testimonial")
    
    # Step 3: Cleanliness & Environment
    streetCleanlinessRating: int = Field(..., ge=1, le=5, description="Street cleanliness rating 1-5")
    trashManagementRating: int = Field(..., ge=1, le=5, description="Trash management rating 1-5")
    parksQualityRating: int = Field(..., ge=1, le=5, description="Parks quality rating 1-5")
    noiseLevel: NoiseLevelType = Field(..., description="Neighborhood noise level")
    environmentalTestimonial: Optional[str] = Field(None, description="Optional environmental testimonial")
    
    # Step 4: Food & Amenities
    groceryStoreAccess: GroceryAccessType = Field(..., description="Grocery store accessibility")
    restaurantVarietyRating: int = Field(..., ge=1, le=5, description="Restaurant variety rating 1-5")
    foodAffordabilityRating: int = Field(..., ge=1, le=5, description="Food affordability rating 1-5")
    farmersMarketAccess: bool = Field(..., description="Access to farmers market")
    foodAccessTestimonial: Optional[str] = Field(None, description="Optional food access testimonial")
    
    # Step 5: Financials & Living Costs
    rentAffordability: AffordabilityType = Field(..., description="Rent/mortgage affordability")
    costOfLiving: CostOfLivingType = Field(..., description="Overall cost of living")
    valueForMoneyRating: int = Field(..., ge=1, le=5, description="Value for money rating 1-5")
    financialTestimonial: Optional[str] = Field(None, description="Optional financial testimonial")
    
    # Step 6: Overall Feedback
    overallRating: int = Field(..., ge=1, le=5, description="Overall neighborhood rating 1-5")
    wouldRecommend: bool = Field(..., description="Would recommend to others")
    biggestStrength: str = Field(..., min_length=10, description="Biggest neighborhood strength")
    areaForImprovement: str = Field(..., min_length=10, description="Area needing improvement")
    additionalComments: Optional[str] = Field(None, description="Optional additional comments")
    
    @field_validator('zipCode')
    @classmethod
    def validate_zip_code(cls, v: str) -> str:
        """Validate zip code is 5 digits"""
        if not re.match(r'^\d{5}$', v):
            raise ValueError('Zip code must be exactly 5 digits')
        return v
    
    model_config = {
        "json_schema_extra": {
            "example": {
                "name": "John Doe",
                "age": 30,
                "borough": "Brooklyn",
                "neighborhood": "Williamsburg",
                "zipCode": "11211",
                "residencyDuration": "1-3 years",
                "safetyRating": 4,
                "timeOfDaySafety": "Both",
                "crimeConcernLevel": "Slightly concerned",
                "policePresenceRating": 4,
                "safetyTestimonial": "Feel safe walking at night",
                "streetCleanlinessRating": 3,
                "trashManagementRating": 3,
                "parksQualityRating": 4,
                "noiseLevel": "Moderate",
                "environmentalTestimonial": "Parks are well maintained",
                "groceryStoreAccess": "Walk < 5min",
                "restaurantVarietyRating": 5,
                "foodAffordabilityRating": 3,
                "farmersMarketAccess": True,
                "foodAccessTestimonial": "Great food options",
                "rentAffordability": "Expensive",
                "costOfLiving": "High",
                "valueForMoneyRating": 3,
                "financialTestimonial": "Worth the cost for the location",
                "overallRating": 4,
                "wouldRecommend": True,
                "biggestStrength": "Amazing restaurants and nightlife",
                "areaForImprovement": "More affordable housing options needed",
                "additionalComments": "Love living here overall"
            }
        }
    }


class SurveyResponse(BaseModel):
    """Survey response schema"""
    id: int
    user_email: str
    name: str
    age: int
    borough: str
    neighborhood: str
    zip_code: str
    residency_duration: str
    safety_rating: int
    time_of_day_safety: str
    crime_concern_level: str
    police_presence_rating: int
    safety_testimonial: str
    street_cleanliness_rating: int
    trash_management_rating: int
    parks_quality_rating: int
    noise_level: str
    environmental_testimonial: str
    grocery_store_access: str
    restaurant_variety_rating: int
    food_affordability_rating: int
    farmers_market_access: bool
    food_access_testimonial: str
    rent_affordability: str
    cost_of_living: str
    value_for_money_rating: int
    financial_testimonial: str
    overall_rating: int
    would_recommend: bool
    biggest_strength: str
    area_for_improvement: str
    additional_comments: str
    created_at: datetime
    message: str = "Survey submitted successfully"
    
    model_config = {
        "from_attributes": True
    }


class ErrorResponse(BaseModel):
    """Error response schema"""
    detail: str
    field_errors: Optional[dict] = None

